### ISSUE TITLE
Bug: does not complain about const use before initializiton

### ISSUE THREAD

**Issue by axkibe** [eslint/eslint#16795]
### Environment

Node version: 16.14.2
npm version: 9.2.0
Local ESLint version: ---
Global ESLint version:  8.29.0
Operating System: debian 11

eslint doesn't see any issue with this broken code:

```
function hello( )
{
	console.log( s );
	const s = 'muhkuh';
	console.log( s );
}

hello( );
```

However node fails with:

```ReferenceError: Cannot access 's' before initialization```

### What parser are you using?

Default (Espree)

### What did you do?

see above

this is my minimal .eslintrc.js

```
module.exports = {
    "extends": "eslint:recommended",
};
```
Actually mine is larger, but thats the minimal sensible that should IMO complain.

### What did you expect to happen?

eslint complaining bitterly about how stupid I am to use a const variable before initializing it.

### What actually happened?

eslinst is happy with my broken code.

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @axkibe, thanks for the issue!

> ```
> function hello( )
> {
> 	console.log( s );
> 	const s = 'muhkuh';
> 	console.log( s );
> }
> 
> hello( );
> ```
> 
> 
>     
>   
> 
> However node fails with:
> 
> 
> `ReferenceError: Cannot access 's' before initialization`

This error can be caught by the [`no-use-before-define`](https://eslint.org/docs/latest/rules/no-use-before-define) rule. Here's a [playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24gaGVsbG8oIClcbntcblx0Y29uc29sZS5sb2coIHMgKTtcblx0Y29uc3QgcyA9ICdtdWhrdWgnO1xuXHRjb25zb2xlLmxvZyggcyApO1xufVxuXG5oZWxsbyggKTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXVzZS1iZWZvcmUtZGVmaW5lIjpbImVycm9yIl19LCJlbnYiOnsiZXM2Ijp0cnVlfX19).

> ```
> module.exports = {
>     "extends": "eslint:recommended",
> };
> ```
> 
> 
>     
>   
> 
> Actually mine is larger, but thats the minimal sensible that should IMO complain.

For the reasoning why `no-use-before-define` rule is not in `eslint:recommended`, please see the discussion in https://github.com/eslint/eslint/issues/11903.

I'm closing this issue since not reporting all possible logic errors that can occur in a code does not constitute a bug in `eslint:recommended`. We're generally in favor of updating `eslint:recommended` with a rule that would report these errors, and as noted in https://github.com/eslint/eslint/issues/11903#issuecomment-817387300 we would welcome a concrete proposal.

**Comment by axkibe** [comment]
I see thanks, my life luck doesn't depend on that feature and I see why it's more problematic than at first glance.

Still I believe with "const" keyword, it never makes sense to refer to it before initialization, but I guess it would make the rule more complicated or the rule would have to been split into different cases.
